HIP/hip-584.md

```diff
@@ -0,0 +1,128 @@
+---
+hip: 584
+title: EVM Module
+author:  Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Mustafa Uzun <mustafa.uzun@limechain.tech>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Review
+created: 2022-09-27
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/586
+updated: 2022-10-04
+---
+
+## Abstract
+
+Encapsulating Hedera smart contract EVM specific logic in a library.
+It would expose interfaces that will allow different implementations.
+Mirror node would build ethereum-related REST API on top of the library. 
+Mirror nodes could use the library to simulate a transaction and the output would be used to 
+populate the return value for the REST API that is invoked. 
+Consensus nodes would also use the library and keep their existing logic by providing specific interface implementations.
+
+## Motivation
+
+Extracting common Hedera smart contract EVM logic in a single place
```

Comment:
The motivation section should be prose not a bullet list.

```suggestion
Currently the EVM code is deeply embedded in consensus node logic.  However ongoing user experience work needs EVM execution to enable features such as cost-free execution of read-only smart contract queries, gas estimation, and transient simulation of read-write operations.

In order to maximize compatibility and reduce defects it would be best to run all these activities from the same codebase.  In order to do so a library needs to be extracted from the consensus logic and the consensus nodes and external programs such as the mirror node.
```